### PR TITLE
Enhance the alarm and event message conversion error logs

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

Make `Alarm` and `Event` message conversion error log messages more verbose by adding payload, topic and error string to the error log message.

For example the `alarm` conversion error message looks as below

```
2022-12-08T05:36:45.031302856Z ERROR tedge-mapper-c8y:messages{ops_dir=Some("/etc/tedge/operations/c8y")}: tedge_mapper::core::converter: Mapping error: Deserialing of alarm message: {
    "message": "Temperature is very high",
    ", received on topic: tedge/alarms/critical/temperature_high, failed due to error: trailing characters at line 4 column 2

```

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_api/src/alarm.rs [hunk 1]
```diff
@@ -42,6 +42,13 @@ pub enum ThinEdgeJsonDeserializerError {
 
     #[error("Unsupported external device ID in topic: {0}")]
     UnsupportedExternalDeviceId(String),
+
+    #[error("Deserialing of alarm message: {payload}, received on topic: {topic}, failed due to error: {error}")]
```
Comment (albinsuresh) [inline crates/core/tedge_api/src/alarm.rs:46]:
```suggestion
    #[error("Parsing of alarm message received on topic: {topic} failed due to error: {error}. Snipped payload: {payload}")]
```

Moving the payload to the end to make the output better.

File: crates/core/tedge_api/src/event.rs [hunk 2]
```diff
@@ -41,6 +41,13 @@ pub mod error {
 
         #[error(transparent)]
         SerdeJsonError(#[from] serde_json::error::Error),
+
+        #[error("Deserialing of event message: {payload}, received on topic: {topic}, failed due to error: {error}")]
```
Comment (albinsuresh) [inline crates/core/tedge_api/src/event.rs:45]: Same comment as above. Move the payload to the end.

File: crates/core/tedge_api/src/alarm.rs (hunk 1 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_api/src/alarm.rs]: Make sense, fixed.

File: crates/core/tedge_api/Cargo.toml [hunk 3]
```diff
@@ -11,6 +11,7 @@ rust-version = "1.58.1"
 clock = { path = "../../common/clock" }
 csv = "1.1"
 download = { path = "../../common/download" }
+error-chain = "0.12.4"
```
Comment (didier-wenzek) [inline crates/core/tedge_api/Cargo.toml:14]: This is unused. I also think this is useless here.

File: crates/core/tedge_mapper/src/c8y/alarm_converter.rs [hunk 4]
```diff
@@ -47,10 +47,16 @@ impl AlarmConverter {
...
-                    input_message.payload_str()?,
-                )?;
+                let mqtt_payload = input_message.payload_str()?;
+                let mqtt_topic = input_message.topic.name.clone();
+                let tedge_alarm =
+                    ThinEdgeAlarm::try_from(&mqtt_topic, mqtt_payload).map_err(|e| {
+                        ThinEdgeJsonDeserializerError::FailedToParseJsonPayload {
+                            topic: mqtt_topic,
+                            error: e.to_string(),
+                            payload: mqtt_payload[0..50].into(),
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/alarm_converter.rs:57]: This panics when the slice end cuts the payload on a non-utf8 boundary.

File: crates/core/tedge_mapper/src/c8y/converter.rs [hunk 5]
```diff
@@ -230,8 +231,17 @@ where
...
-        let tedge_event = ThinEdgeEvent::try_from(&input.topic.name, input.payload_str()?)?;
+        let mqtt_payload = input.payload_str()?;
+        let mqtt_topic = input.topic.name.clone();
+
+        let tedge_event = ThinEdgeEvent::try_from(&input.topic.name, input.payload_str()?)
+            .map_err(
+                |e| ThinEdgeJsonDeserializerError::FailedToParseJsonPayload {
+                    topic: mqtt_topic,
+                    error: e.to_string(),
+                    payload: mqtt_payload[0..50].into(),
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/converter.rs:242]: This panics when the slice end cuts the payload on a non-utf8 boundary.

File: crates/core/tedge_mapper/src/c8y/converter.rs [hunk 6]
```diff
@@ -230,8 +231,17 @@ where
         input: &Message,
     ) -> Result<Vec<Message>, ConversionError> {
         let mut messages = Vec::new();
-
-        let tedge_event = ThinEdgeEvent::try_from(&input.topic.name, input.payload_str()?)?;
+        let mqtt_payload = input.payload_str()?;
+        let mqtt_topic = input.topic.name.clone();
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/converter.rs:235]: Minor. I would avoid to clone the topic and to extract the payload twice even when there is no error.

File: crates/core/tedge_mapper/src/c8y/alarm_converter.rs [hunk 7]
```diff
@@ -47,10 +47,16 @@ impl AlarmConverter {
             }
             Self::Synced => {
                 //Regular conversion phase
-                let tedge_alarm = ThinEdgeAlarm::try_from(
-                    input_message.topic.name.as_str(),
-                    input_message.payload_str()?,
-                )?;
+                let mqtt_payload = input_message.payload_str()?;
+                let mqtt_topic = input_message.topic.name.clone();
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y/alarm_converter.rs:51]: Minor. I would avoid to clone the topic and to extract the payload twice even when there is no error.

**Review by didier-wenzek — CHANGES_REQUESTED** [review]
The code panics if the payload excerpt is not aligned on utf-8 boundaries.

File: crates/core/tedge_mapper/src/c8y/alarm_converter.rs (hunk 4 above)
Comment (didier-wenzek) [reply crates/core/tedge_mapper/src/c8y/alarm_converter.rs:57]: Use: `mqtt_payload.chars.take(50).collect()`.

File: crates/core/tedge_api/Cargo.toml (hunk 3 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_api/Cargo.toml]: Fixed

File: crates/core/tedge_mapper/src/c8y/alarm_converter.rs (hunk 4 above)
Comment (PradeepKiruvale) [reply crates/core/tedge_mapper/src/c8y/alarm_converter.rs]: Fixed

File: crates/core/tedge_api/src/event.rs [hunk 8]
```diff
@@ -48,6 +48,9 @@ pub mod error {
             error: String,
             payload: String,
         },
+
+        #[error("Failed to parse the payload to string that was received on topic: {topic}, due to error: {error}.")]
+        FailedToParsePayloadToString { topic: String, error: String },
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/event.rs:53]: This error message is unclear for a non rust developer.

```suggestion
        #[error("Failed to parse as an UTF-8 string the payload received on topic: {topic}, due to error: {error}.")]
        FailedToParsePayloadToString { topic: String, error: String },
```

File: crates/core/tedge_api/src/alarm.rs [hunk 9]
```diff
@@ -49,6 +49,9 @@ pub enum ThinEdgeJsonDeserializerError {
         error: String,
         payload: String,
     },
+
+    #[error("Parsing of alarm message to string received on topic: {topic} failed due to error: {error}")]
+    FailedToParsePayloadToString { topic: String, error: String },
```
Comment (didier-wenzek) [inline crates/core/tedge_api/src/alarm.rs:54]: This error message is unclear for a non rust developer.

```suggestion
    #[error("Failed to parse as an UTF-8 string the payload received on topic: {topic}, due to error: {error}.")]
    FailedToParsePayloadToString { topic: String, error: String },
```

Comment (PradeepKiruvale) [reply crates/core/tedge_api/src/alarm.rs]: updated the message

File: crates/core/tedge_api/src/event.rs [hunk 10]
```diff
@@ -41,6 +41,16 @@ pub mod error {
 
         #[error(transparent)]
         SerdeJsonError(#[from] serde_json::error::Error),
+
+        #[error("Parsing of alarm message received on topic: {topic} failed due to error: {error}. Snipped payload: {payload}")]
```
[the remark this answers is not in the document]
Comment (PradeepKiruvale) [reply crates/core/tedge_api/src/event.rs:45]: Changed the error message as suggested by Didier below.

**Review by didier-wenzek — APPROVED** [review]
Approved. Thanks for this improvement.

**Comment by rina23q** [comment]
@PradeepKiruvale You need to write how to reproduce the error messages that you fixed if you think this PR needs to be verified by QA.

**Comment by PradeepKiruvale** [comment]
> @PradeepKiruvale You need to write how to reproduce the error messages that you fixed if you think this PR needs to be verified by QA.

The steps are added here:

https://github.com/thin-edge/thin-edge.io/issues/1627

**Pull request merged**